### ISSUE TITLE
Review RSA operation with large modulus and/or large public exponents

### ISSUE THREAD

**Issue by AnnieYousar** [openssl/openssl#6130]
Prepare two RSA keys by the following commands (keep calm for the second):
```
openssl genpkey -algorithm rsa -pkeyopt rsa_keygen_bits:4096  -pkeyopt rsa_keygen_pubexp:0x010000000000000001 -out rsa04096.key
openssl genpkey -algorithm rsa -pkeyopt rsa_keygen_bits:17000 -pkeyopt rsa_keygen_pubexp:0x1000000000000001 -out rsa17000.key
```
Then try to use these keys for encryption and signing. Remember that for encryption you use the public exponent, and for signing the private.
```
openssl rsautl -encrypt -in test.txt -out test.enc -inkey rsa04096.key
openssl rsautl -sign -in test.txt -out test.sig -inkey rsa04096.key
openssl rsautl -encrypt -in test.txt -out test.enc -inkey rsa17000.key
openssl rsautl -sign -in test.txt -out test.sig -inkey rsa17000.key
```
Signing passes, encryption not. With 4096 key due to "bad e value", and the 17000 key modulus is too large. OK, encryption failed. Try to verify the signature file `test.sig`:
```
openssl rsautl -verify -in test.sig -out test.ver -inkey rsa04096.key
openssl rsautl -verify -in test.sig -out test.ver -inkey rsa17000.key
```  
Both fail, again because of "bad e value" and "modulus too large".
Strange, isn't it?
Explanation:
The public (not the private) key operations are restricted in `include/openssl/rsa.h` by 
```
define OPENSSL_RSA_MAX_MODULUS_BITS   16384
define OPENSSL_RSA_SMALL_MODULUS_BITS 3072
/* exponent limit enforced for "large" modulus only */
define OPENSSL_RSA_MAX_PUBEXP_BITS    64
```
Therefore signing with an indeed large modulus is fine, encryption not. If the modulus length is greater the 3072, then additional the bit length of the exponent is restricted.

1. If anybody knows, please explain why 16385 is too large for OpenSSL.

2. For the public exponent there are some reasons to restrict the bit length. A small length of the public key ensures a large length of the private exponent. Small private exponents are risky. Therefore if you see a small public exponent, you can be sure that the corresponding private exponent is large enough.
Usually (NIST did it since FIPS 186-3) the bit length of the public exponent should be less than 256 (for modulus length not smaller 2048). For a smaller modulus length I guess, it should be less than 64.

**Comment by kroeckx** [comment]
> 1. If anybody knows, please explain why 16385 is too large for OpenSSL.

I think it's mostly because it becomes really slow and doesn't really add something.

> 2. For the public exponent there are some reasons to restrict the bit length.

The cabforum baseline requirements currently says:
```
The CA SHALL confirm that the value of the public exponent is an odd number equal to 3 or more. 
Additionally, the public exponent SHOULD be in the range between 2^16+1 and 2^256-1. The modulus SHOULD 
also have the following characteristics: an odd number, not the power of a prime, and have no factors smaller 
than 752. [Source: Section 5.3.3, NIST SP 800-89]
```

**Comment by kroeckx** [comment]
I'm also not sure what you want us to do about it.

**Comment by AnnieYousar** [comment]
I would like to discuss the following ideas:
1. Remove the length restriction `define OPENSSL_RSA_MAX_MODULUS_BITS   16384`
The key generation may be slow. The secret key operation (17000 bit key) on a quad core PC lasts 8 sec (not really slow), the public key operation 0.13 sec. You could say that the secret key operation is slow, but  this length restriction applies to the pubkey operation only. It makes no sense, that OpenSSL generates long keys but can't use them.
2. Adjust the overall exponent length restriction `define OPENSSL_RSA_MAX_PUBEXP_BITS    64`
Set it to 256 bit and do it for arbitrary key lengths (as NIST recommends).
3. Remove the lower bound `define OPENSSL_RSA_SMALL_MODULUS_BITS 3072`  on the key length for the exponent length restriction.
4. At least warn if the public exponent is outside the NIST range (2^16,2^256).

**Comment by AnnieYousar** [comment]
The proposed public key length restriction (2.) prevents the usage of the following public RSA key
```
-----BEGIN PUBLIC KEY-----
MIIBGzANBgkqhkiG9w0BAQEFAAOCAQgAMIIBAwKBgQDI0rl3Jrbj+43Uki58yYHx
fcnZPIfaesoY4eFlH5ClIfTKq4iEoXmzTPxlBeanR16DZYldeT0iW/c4+k6HrqA3
zh2sMrgGapv/sGaaBexMkTOyPgdN6VggDw6A9Immtc6CFXiB+RMlxTyvoGUB4HSU
HVG8HFVxDb4xhoEYV/O59QJ9ARxukLZ3/5exYd2x06yOYJiuoKDhysp6GaLdWnuK
j4DAcjpRhMrHE0B1Uc5rvSpsZVRY4/CFwdWBCc+sKfi1gL1tH5VGgnlV+IrvNgCE
tVjOywrB3cYAFa2TOyivhoOSt9vrhDDYZrwYGr+8RsgM18peHrpHHFWPgxRsopc=
-----END PUBLIC KEY-----
```
which is completely broken. Try, e.g., with this key the command
`openssl rsautl -encrypt -in test.txt -pubin -inkey rsa1024.pub -out test.enc`
and enjoy the `xxd` of the output.

**Comment by kroeckx** [comment]
The output contains the plain text input with that example ...

**Comment by AnnieYousar** [comment]
Yeah, the public exponent is equivalent to 1, but nobody can see this from the key data.

**Comment by AnnieYousar** [comment]
The key is weak, because `e \equiv d \equiv 1 \bmod LCD(p-1,q-1)`. In this case either e is 1 or its bit length is at least half the key size. A weak key (with e \equiv 1) could be checked out easily, if the bit size of the exponent is small. 
The same approach works for keys with small private exponents d. They are weak, but this can not occur is the public key is small enough.

**Comment by richsalz** [comment]
This does not seem like something we should hold the 1.1.1 release for (we could fix in 1.1.1a for example).  Moving this out.  Please comment if you disagree.

**Comment by AnnieYousar** [comment]
It's fine with me. Note that NIST in FIPS 186-4 (https://nvlpubs.nist.gov/nistpubs/fips/nist.fips.186-4.pdf, p. 52) requires (i.e. SHALL) that "the exponent `e` shall  be an odd positive integer such that `2^16 < e < 2^256`. This excludes such silly keys as given above.

**Comment by kroeckx** [comment]
It wasn't a request to close this, it's just not a blocker for the 1.1.1 release. So I'm reopening it.

**Comment by AnnieYousar** [comment]
Sorry, I mispressed the button ;-)

**Comment by briansmith** [comment]
There needs to be some limit on the sizes because internally the RSA code uses `alloca()` based on the input sizes. Note that these `alloca()` are done implicitly in the assembly language code so you can't necessarily find them by grepping for `alloca` (though many have comments that include "alloca").

IMO it would be better to pull the `alloca()` out of the assembly language code so that they are more readily apparent and so we can more easily review whether the current limits could be loosened or whether we need stricter limitations and/or whether we need to enforce the limitations in more places (e.g. enforce limits in the BN code itself, rather than just in RSA, et al.).

**Comment by davidben** [comment]
> IMO it would be better to pull the alloca() out of the assembly language code so that they are more readily apparent and so we can more easily review whether the current limits could be loosened or whether we need stricter limitations and/or whether we need to enforce the limitations in more places (e.g. enforce limits in the BN code itself, rather than just in RSA, et al.).

Agreed. I've filed https://github.com/openssl/openssl/issues/6385 for this.

**Comment by AnnieYousar** [comment]
I see, that the length restriction for the modulus is a different issue. 
But the length restrictions for the exponent `2^16 < e < 2^256` is independent from this.

**Comment by voidastro4** [comment]
Increasing RSA limit (16384) and DH limit (10000) to arbitrary values would if nothing else buy years as a hypothetical quantum computer scales the number of qubits it has. And there is a chance a quantum computer simply cannot scale past a certain point.

Having DH params for 32768, 65536, 131072, 262144, etc would be useful for some long term/paranoid applications where the degraded performance is of no relevance.

Arbitrary limits of on user choice hardcoded into the code serve no real purpose.

**Comment by davidben** [comment]
These are DoS limits. These operations scale cubically with the size of the key. Ideally RSA-2048 and RSA-4096 would simply have been distinct algorithms rather than points in a single variable-size algorithm, but that wasn't how things shook out, so we've got to have these ad hoc limits.
